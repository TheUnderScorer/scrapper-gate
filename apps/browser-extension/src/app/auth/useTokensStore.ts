import {
  makeUseTokenStore,
  PersistentTokensStorage,
} from '@scrapper-gate/frontend/domain/auth';
import browser from 'webextension-polyfill';

const storage: PersistentTokensStorage = {
  set: async (tokens) => {
    await browser.storage.local.set({
      tokens: tokens ?? null,
    });
  },
  get: async () =>
    browser.storage.local.get('tokens').then((result) => result.tokens),
};

export const useTokensStore = makeUseTokenStore(storage);
