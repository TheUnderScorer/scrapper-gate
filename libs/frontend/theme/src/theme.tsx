import { Delete, PlayArrow } from '@mui/icons-material';
import { createTheme, SimplePaletteColorOptions } from '@mui/material/styles';
import React from 'react';
import { conditionalStopPropagation } from './conditionalStopPropagation';
import { palette } from './palette';

interface ThemeParams {
  isContent?: boolean;
  container?: HTMLElement;
  htmlFontSize: number;
}

export const getTheme = ({ isContent, container, htmlFontSize }: ThemeParams) =>
  createTheme({
    palette,
    components: {
      MuiButton: {
        defaultProps: {
          disableElevation: true,
        },
        styleOverrides: {
          containedPrimary: {
            background: palette.gradients.primaryMainToDark,
            color: (palette.primary as SimplePaletteColorOptions).contrastText,
          },
        },
      },
      MuiFab: {
        styleOverrides: {
          primary: {
            '&:not[disabled]': {
              background: palette.gradients.primaryMainToDark,
            },
          },
        },
      },
      MuiPopover: {
        defaultProps: {
          disableEnforceFocus: isContent,
          container,
        },
      },
      MuiTooltip: {
        defaultProps: {
          PopperProps: {
            container,
          },
        },
      },
      MuiModal: {
        defaultProps: {
          disableEnforceFocus: isContent,
          container,
        },
      },
      MuiMenu: {
        defaultProps: {
          container,
          style: {
            pointerEvents: 'all',
          },
        },
      },
      MuiAccordion: {
        variants: [
          {
            props: {
              variant: 'transparent',
            },
            style: {
              background: 'transparent',
              border: 'none',
            },
          },
        ],
      },
      MuiDialog: {
        defaultProps: {
          container,
          disableEnforceFocus: isContent,
          disableAutoFocus: isContent,
          disableRestoreFocus: isContent,
          BackdropProps: {
            style: {
              pointerEvents: 'all',
            },
          },
          PaperProps: {
            style: {
              pointerEvents: 'all',
            },
          },
        },
      },
      MuiInput: {
        defaultProps: {
          onKeyDown: conditionalStopPropagation,
          onKeyUp: conditionalStopPropagation,
          onKeyPress: conditionalStopPropagation,
        },
      },
      MuiTextField: {
        defaultProps: {
          onKeyDown: conditionalStopPropagation,
          onKeyUp: conditionalStopPropagation,
          onKeyPress: conditionalStopPropagation,
        },
      },
    },
    typography: {
      fontSize: 13,
      htmlFontSize,
    },
    zIndex: {
      modal: 1400,
    },
    emojis: {
      empty: 'üòØ',
      error: 'üòµ',
      success: 'üòÅ',
    },
    icons: {
      run: <PlayArrow />,
      delete: <Delete />,
    },
  });
