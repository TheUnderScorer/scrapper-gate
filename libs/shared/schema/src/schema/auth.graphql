# REST - don't remove

input LoginInput {
  username: String!
  password: String!
}

type LoginResponse {
  accessToken: String!
  refreshToken: String!
}

input ForgotPasswordInput {
  username: String!
}

type ForgotPasswordResponse {
  error: String
  stack: String
}

input ResetPasswordInput {
  newPassword: String!
}

type ResetPasswordResponse {
  error: String
  stack: String
}

type IsAutenthicatedResponse {
  isAutenthicated: Boolean
}

# END REST

extend type Mutation {
  #REST don't remove
  login(input: LoginInput!): LoginResponse
  forgotPassword(input: ForgotPasswordInput): ForgotPasswordResponse
  resetPassword(
    input: ResetPasswordInput
    token: String!
  ): ResetPasswordResponse
  #END REST
}

extend type Query {
  isAutenthicated: IsAutenthicatedResponse
}
