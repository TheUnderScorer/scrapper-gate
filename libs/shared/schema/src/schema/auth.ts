import gql from 'graphql-tag';

export const authSchema = gql`
  type AuthTokens {
    accessToken: String!
    refreshToken: String!
  }
`;
