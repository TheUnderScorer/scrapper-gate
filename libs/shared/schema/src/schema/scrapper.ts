import gql from 'graphql-tag';

export const scrapperSchema = gql`
  type Scrapper implements BaseEntity & CreatedBy {
    id: ID!
    createdAt: Date!
    updatedAt: Date!
    isRunning: Boolean
    name: String
    state: RunState
    createdBy: User
    deletedAt: Date
    steps: [ScrapperStep!]
  }

  type ScrapperQueryResult {
    total: Int!
    items: [Scrapper!]
  }

  type ScrapperStep implements BaseEntity & CreatedBy {
    id: ID!
    createdAt: Date!
    updatedAt: Date!
    deletedAt: Date
    createdBy: User
    goBackSteps: Int
    nextStep: ScrapperStep
    mouseButton: MouseButton
    url: Url
    navigateToUrl: Url
    reloadDelay: Float
    typeDelay: Float
    useUrlFromPreviousStep: Boolean
    action: ScrapperAction
    selectors: [Selector!]
  }

  input ScrapperStepInput {
    id: ID
    goBackSteps: Int
    nextStepId: ID
    mouseButton: MouseButton
    url: Url
    navigateToUrl: Url
    reloadDelay: Float
    typeDelay: Float
    useUrlFromPreviousStep: Boolean
    action: ScrapperAction
    selectors: [SelectorInput!]
  }

  enum ScrapperAction {
    Click
    Condition
    GoBack
    NavigateTo
    ReadText
    ReloadPage
    Type
  }

  enum MouseButton {
    Left
    Right
    Middle
  }

  input CreateScrapperInput {
    name: String
  }

  extend type Mutation {
    createScrapper(input: CreateScrapperInput): Scrapper! @auth
  }

  extend type Query {
    getMyScrappers(pagination: Pagination, order: Order): ScrapperQueryResult!
      @auth
  }
`;
