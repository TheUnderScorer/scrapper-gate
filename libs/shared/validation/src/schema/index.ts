export * from './conditionalRules/HtmlConditionalRule.schema';
export * from './conditionalRules/VariableConditionalRule.schema';
export * from './conditionalRules/DateRule.schema';
export * from './scrapper/ScrapperInput.schema';
export * from './scrapper/ScrapperStepInput.schema';
export * from './scrapper/StartScrapperInput.schema';
export * from './scrapper/ScrapperConditionalRuleGroupInput.schema';
export * from './scrapper/ScrapperRunnerMessagePayload.schema';
export * from './scrapper/ScrapperRunSettingsInput.schema';
export * from './variables/VariableInput.schema';
export * from './DurationInput.schema';
export * from './LoginInput.schema';
export * from './PositionInput.schema';
export * from './Selector.schema';
export * from './scrapper/CreateScrapperInput.schema';
