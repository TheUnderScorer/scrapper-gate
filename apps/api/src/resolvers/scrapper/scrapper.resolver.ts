import { Resolvers } from '@scrapper-gate/shared/schema';
import { BaseApolloContext } from '@scrapper-gate/backend/server';
import {
  CreateScrapperCommand,
  GetScrappersByUserQuery,
} from '@scrapper-gate/backend/domain/scrapper';
import { UserModel } from '@scrapper-gate/backend/domain/user';

export const scrapperResolver = (): Resolvers<
  BaseApolloContext<UserModel>
> => ({
  Query: {
    getMyScrappers: (_, args, ctx) =>
      ctx.unitOfWork.run(({ queriesBus }) =>
        queriesBus.query(
          new GetScrappersByUserQuery({
            ...args,
            userId: ctx.user.id,
          })
        )
      ),
  },
  Mutation: {
    createScrapper: (_, args, ctx) =>
      ctx.unitOfWork.run(({ commandsBus }) =>
        commandsBus.execute(
          new CreateScrapperCommand({
            input: args.input,
            user: ctx.user,
          })
        )
      ),
  },
  Scrapper: {
    name: (root) => root.name ?? 'Unnamed scrapper',
  },
});
