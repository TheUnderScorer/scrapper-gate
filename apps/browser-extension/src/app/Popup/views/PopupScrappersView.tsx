import React from 'react';
import { MyScrappersList } from '@scrapper-gate/frontend/domain/scrapper';
import { Centered, Image, useAsset } from '@scrapper-gate/frontend/ui';
import { Fab, makeStyles, Typography } from '@material-ui/core';
import { useCreateScrapperExtension } from '../hooks/useCreateScrapperExtension';

const useStyles = makeStyles((theme) => ({
  empty: {
    backgroundColor: theme.palette.primary.light,
  },
  text: {
    marginTop: theme.spacing(4),
    marginBottom: theme.spacing(3),
  },
}));

export const PopupScrappersView = () => {
  const classes = useStyles();

  const { asset, alt } = useAsset('notFoundSolid');

  const [createScrapper] = useCreateScrapperExtension();

  return (
    <MyScrappersList
      onCreate={() => createScrapper()}
      emptyContent={
        <Centered direction="column" className={classes.empty}>
          <Image alt={alt} src={asset} />
          <Typography className={classes.text} variant="subtitle1">
            No scrappers found.
          </Typography>
          <Fab
            onClick={() => createScrapper()}
            color="primary"
            variant="extended"
          >
            Create first scrapper
          </Fab>
        </Centered>
      }
    />
  );
};
