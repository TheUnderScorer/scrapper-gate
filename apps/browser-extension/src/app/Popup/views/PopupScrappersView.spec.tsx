import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { PopupScrappersView } from './PopupScrappersView';
import { act, render, waitFor } from '@testing-library/react';
import { Tabs } from 'webextension-polyfill-ts';
import { MyScrappersDocument } from '@scrapper-gate/frontend/schema';
import {
  MyScrappersQuery,
  MyScrappersQueryVariables,
} from '@scrapper-gate/shared/schema';
import { initialPagination } from '@scrapper-gate/frontend/ui';
import { createMockScrapper } from '@scrapper-gate/shared/domain/scrapper';
import userEvent from '@testing-library/user-event';
import { mockEvent, MockzillaEventOf } from 'mockzilla-webextension';
import { useTokensStore } from '@scrapper-gate/frontend/domain/auth';
import { browserExtensionRoutes } from '@scrapper-gate/shared/routing';
import {
  MessageTypes,
  StoredRoute,
} from '../../../extension/browser/communication/types';
import { ThemeProvider } from '@scrapper-gate/frontend/theme';

const scrapper = createMockScrapper();
scrapper.name = 'Test';

const defaultMocks: MockedResponse[] = [
  {
    request: {
      query: MyScrappersDocument,
      variables: {
        pagination: initialPagination,
        order: undefined,
      } as MyScrappersQueryVariables,
    },
    result: {
      data: {
        getMyScrappers: {
          total: 1,
          items: [scrapper],
        },
      } as MyScrappersQuery,
    },
  },
];

const mountCmp = (mocks = defaultMocks) => {
  return render(
    <MockedProvider mocks={mocks}>
      <ThemeProvider>
        <PopupScrappersView fetchPolicy="network-only" />
      </ThemeProvider>
    </MockedProvider>
  );
};

const waitForListItem = async (container: HTMLElement) => {
  await waitFor(
    () => {
      const listItem = container.querySelector('.scrapper-list-item');

      return expect(listItem).not.toBeNull();
    },
    {
      interval: 250,
      timeout: 3000,
    }
  );
};

const mockTab = {
  id: 1,
  windowId: 1,
};

const tokens = {
  accessToken: '#access',
  refreshToken: '#refresh',
};

function defaultStoreMock() {
  mockBrowser.storage.local.get.mock(async () => ({
    contentRoutes: {
      [mockTab.id]: {},
    },
  }));
}

describe('Popup scrappers view', () => {
  let onMessage: MockzillaEventOf<typeof mockBrowser.runtime.onMessage>;

  beforeEach(() => {
    onMessage = mockEvent(mockBrowser.runtime.onMessage);
    mockBrowser.tabs.query.mock(async () => [mockTab as Tabs.Tab]);

    useTokensStore.setState({
      tokens,
    });
  });

  it('should render without crashing', async () => {
    defaultStoreMock();

    const cmp = mountCmp();

    await waitForListItem(cmp.container);

    expect(cmp).toMatchSnapshot();
  });

  it('should send message to content after clicking scrapper', async () => {
    defaultStoreMock();

    const sendMessage = jest.fn();
    mockBrowser.runtime.sendMessage.mock((...args) => sendMessage(...args));

    const cmp = mountCmp();

    await waitForListItem(cmp.container);

    const listItem = cmp.container.querySelector('.scrapper-list-item');

    act(() => {
      userEvent.click(listItem);
    });

    expect(sendMessage).toHaveBeenCalledWith({
      payload: {
        tokens,
        visible: true,
        path: browserExtensionRoutes.content.scrapper({
          scrapperId: scrapper.id,
          drawerOpen: true,
        }),
      },
      type: MessageTypes.ToggleContent,
    });
  });

  it('should send message that closes content if user clicks active scrapper', async () => {
    const sendMessage = jest.fn();
    mockBrowser.runtime.sendMessage.mock((...args) => sendMessage(...args));

    mockBrowser.storage.local.get.mock(async () => ({
      contentRoutes: {
        [mockTab.id]: {
          pathname: browserExtensionRoutes.content.scrapper({
            scrapperId: scrapper.id,
            drawerOpen: true,
          }),
        } as StoredRoute,
      },
    }));

    const cmp = mountCmp();

    await waitForListItem(cmp.container);

    const listItem = cmp.container.querySelector('.scrapper-list-item');

    act(() => {
      userEvent.click(listItem);
    });

    expect(sendMessage).toHaveBeenCalledWith({
      payload: {
        tokens,
        visible: false,
        path: '/',
      },
      type: MessageTypes.ToggleContent,
    });
  });
});
