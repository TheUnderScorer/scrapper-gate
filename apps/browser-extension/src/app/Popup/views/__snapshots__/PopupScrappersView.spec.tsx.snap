// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup scrappers view should render without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-wrap-xs-nowrap makeStyles-container-5 layout css-19kb62l-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item makeStyles-body-6 makeStyles-body-7 layout-body css-10texkh-MuiGrid-root"
          style="height: calc(100% - 0px); flex: 1;"
        >
          <ul
            class="MuiList-root MuiList-padding makeStyles-list-4 controlled-list css-h4y409-MuiList-root"
            id="my_scrappers_list"
          >
            <li
              class="MuiListItem-container css-oyakqy-MuiListItem-container"
            >
              <div
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button MuiListItem-secondaryAction scrapper-list-item css-1cdun9u-MuiButtonBase-root-MuiListItem-root"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                  >
                    Test
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </div>
              <div
                class="MuiListItemSecondaryAction-root css-518kzi-MuiListItemSecondaryAction-root"
              >
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1uaxfgw-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label css-4jkopv-MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="MoreVertSharpIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </li>
          </ul>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-primary makeStyles-fab-3 css-mv5a4t-MuiButtonBase-root-MuiFab-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiFab-label css-110lfi1-MuiFab-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      aria-hidden="true"
      class="MuiModal-root MuiModal-hidden MuiPopover-root MuiMenu-root css-6f6ff7-MuiModal-root-MuiPopover-root-MuiMenu-root"
      role="presentation"
      style="pointer-events: all;"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root MuiBackdrop-invisible css-1qjps05-MuiBackdrop-root"
        style="opacity: 0; visibility: hidden;"
      />
      <div
        data-test="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiPopover-paper css-1poimk-MuiPaper-root-MuiMenu-paper-MuiPaper-root-MuiPopover-paper"
        style="opacity: 0; transform: scale(0.75, 0.5625); visibility: hidden;"
        tabindex="-1"
      >
        <ul
          class="MuiList-root MuiList-padding MuiMenu-list css-6hp17o-MuiList-root-MuiMenu-list"
          role="menu"
          tabindex="-1"
        >
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button MuiMenuItem-root css-4ssu7w-MuiButtonBase-root-MuiListItem-root-MuiMenuItem-root"
            id="delete_scrapper"
            role="menuitem"
            tabindex="-1"
          >
            <div
              class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="DeleteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </div>
            <div
              class="MuiListItemText-root css-tlelie-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                Delete scrapper
              </span>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
        </ul>
      </div>
      <div
        data-test="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-wrap-xs-nowrap makeStyles-container-5 layout css-19kb62l-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item makeStyles-body-6 makeStyles-body-7 layout-body css-10texkh-MuiGrid-root"
        style="height: calc(100% - 0px); flex: 1;"
      >
        <ul
          class="MuiList-root MuiList-padding makeStyles-list-4 controlled-list css-h4y409-MuiList-root"
          id="my_scrappers_list"
        >
          <li
            class="MuiListItem-container css-oyakqy-MuiListItem-container"
          >
            <div
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button MuiListItem-secondaryAction scrapper-list-item css-1cdun9u-MuiButtonBase-root-MuiListItem-root"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemText-root css-tlelie-MuiListItemText-root"
              >
                <span
                  class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                >
                  Test
                </span>
              </div>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </div>
            <div
              class="MuiListItemSecondaryAction-root css-518kzi-MuiListItemSecondaryAction-root"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1uaxfgw-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label css-4jkopv-MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="MoreVertSharpIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </li>
        </ul>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-primary makeStyles-fab-3 css-mv5a4t-MuiButtonBase-root-MuiFab-root"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiFab-label css-110lfi1-MuiFab-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="AddIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
